storage/remote/queue_manager.go

```diff
@@ -468,18 +470,46 @@ func (t *QueueManager) updateShardsLoop() {
 	for {
 		select {
 		case <-ticker.C:
-			t.calculateDesiredShards()
+			desiredShards := t.calculateDesiredShards()
+			if desiredShards == t.numShards {
+				continue
+			}
+			// Resharding can take some time, and we want this loop
+			// to stay close to shardUpdateDuration.
+			select {
+			case t.reshardChan <- desiredShards:
+				level.Info(t.logger).Log("msg", "Remote storage resharding", "from", t.numShards, "to", numShards)
+				t.numShards = desiredShards
+			default:
+				level.Info(t.logger).Log("msg", "Currently resharding, skipping.")
+			}
 		case <-t.quit:
 			return
 		}
 	}
 }
 
-func (t *QueueManager) calculateDesiredShards() {
+// calculateDesiredShards returns the number of desired shards, which will be
+// the current QueueManager.numShards if resharding should not occur for reasons
+// outlined in this functions implementation. It is up to the caller to reshard, or not,
+// based on the return value.
+func (t *QueueManager) calculateDesiredShards() int {
 	t.samplesOut.tick()
 	t.samplesDropped.tick()
 	t.samplesOutDuration.tick()
 
+	// We shouldn't reshard if Prometheus hasn't been able to send to the
+	// remote endpoint successfully within some period of time.
+	now := time.Now().Unix()
+	threshold := int64(time.Duration(2 * t.cfg.BatchSendDeadline).Seconds())
+	t.lastSendTimestampLock.RLock()
+	diff := now - t.lastSendTimestamp
+	t.lastSendTimestampLock.RUnlock()
+	if diff > threshold {
+		level.Warn(t.logger).Log("msg", "Skipping resharding, last successful send was beyond threshold")
+		return t.numShards
```

Comment:
By breaking out of the loop before trying to add to the integral accumulator you are not be accounting for tons of pending samples when Prometheus starts being able to send again. This will cause future shard calculations to under provision the number of necessary shards.